### ISSUE TITLE
Shouldn't SWIG_V8_ConvertInstancePtr call SWIG_TypeCast if types are not equal?

### ISSUE THREAD

**Issue by murillo128** [swig/swig#1498]
I have the current code snippets:

```
%nodefaultctor AudioInput;
%nodefaultdtor AudioInput;
struct  AudioInput {};

%nodefaultctor AudioOutput;
%nodefaultdtor AudioOutput;
struct  AudioOutput {};

struct AudioPipe : 
	public AudioInput,
	public AudioOutput
{
	AudioPipe(int rate);
};


struct AudioDecoderFacade
{
	int Start();
	void AddAudioOuput(AudioOutput* ouput);
	void RemoveAudioOutput(AudioOutput* ouput);
	bool SetIncoming(RTPIncomingMediaStream* incoming);
	int Stop();
};
```

which generates the following code for the `AddAudioOuput` method:

```

static SwigV8ReturnValue _wrap_AudioDecoderFacade_AddAudioOuput(const SwigV8Arguments &args) {
  SWIGV8_HANDLESCOPE();
  
  v8::Handle<v8::Value> jsresult;
  AudioDecoderFacade *arg1 = (AudioDecoderFacade *) 0 ;
  AudioOutput *arg2 = (AudioOutput *) 0 ;
  void *argp1 = 0 ;
  int res1 = 0 ;
  void *argp2 = 0 ;
  int res2 = 0 ;
  
  if(args.Length() != 1) SWIG_exception_fail(SWIG_ERROR, "Illegal number of arguments for _wrap_AudioDecoderFacade_AddAudioOuput.");
  
  res1 = SWIG_ConvertPtr(args.Holder(), &argp1,SWIGTYPE_p_AudioDecoderFacade, 0 |  0 );
  if (!SWIG_IsOK(res1)) {
    SWIG_exception_fail(SWIG_ArgError(res1), "in method '" "AudioDecoderFacade_AddAudioOuput" "', argument " "1"" of type '" "AudioDecoderFacade *""'"); 
  }
  arg1 = (AudioDecoderFacade *)(argp1);
  res2 = SWIG_ConvertPtr(args[0], &argp2,SWIGTYPE_p_AudioOutput, 0 |  0 );
  if (!SWIG_IsOK(res2)) {
    SWIG_exception_fail(SWIG_ArgError(res2), "in method '" "AudioDecoderFacade_AddAudioOuput" "', argument " "2"" of type '" "AudioOutput *""'"); 
  }
  arg2 = (AudioOutput *)(argp2);
  (arg1)->AddAudioOuput(arg2);
  jsresult = SWIGV8_UNDEFINED();
  
  
  
  SWIGV8_RETURN(jsresult);
  
  goto fail;
fail:
  SWIGV8_RETURN(SWIGV8_UNDEFINED());
}
```
So if I create an `AudioPipe` object (which public inherits from `AudioOuput` and from `AudioInput`) then no cast is called, and directly a cast from the `void*` pointer to `(AudioOutput*)` is done.
 
Shouldn't `SWIG_ConvertPtr` call `SWIG_TypeCast` after checking that the type cast is correct?

In `SWIG_V8_ConvertInstancePtr`

```
SWIGRUNTIME int SWIG_V8_ConvertInstancePtr(v8::Handle<v8::Object> objRef, void **ptr, swig_type_info *info, int flags) {
  SWIGV8_HANDLESCOPE();

  if(objRef->InternalFieldCount() < 1) return SWIG_ERROR;

#if (V8_MAJOR_VERSION-0) < 4 && (SWIG_V8_VERSION < 0x031511)
  v8::Handle<v8::Value> cdataRef = objRef->GetInternalField(0);
  SWIGV8_Proxy *cdata = static_cast<SWIGV8_Proxy *>(v8::External::Unwrap(cdataRef));
#else
  SWIGV8_Proxy *cdata = static_cast<SWIGV8_Proxy *>(objRef->GetAlignedPointerFromInternalField(0));
#endif

  if(cdata == NULL) {
    return SWIG_ERROR;
  }
  if(cdata->info != info) {
    swig_cast_info *tc = SWIG_TypeCheckStruct(cdata->info, info);
    if (!tc && cdata->info->name) {
      tc = SWIG_TypeCheck(cdata->info->name, info);
    }
    bool type_valid = tc != 0;
    if(!type_valid) {
      return SWIG_TypeError;
    }
    ///<--- Do conversion here! 
  }
  *ptr = cdata->swigCObject;
  if(flags & SWIG_POINTER_DISOWN) {
    cdata->swigCMemOwn = false;
  }
  return SWIG_OK;
}

```

**Comment by murillo128** [comment]
@lissyx @wsfulton could you take a look at this? 

Without the SWIG_TypeCast I am getting calls to the wrong vtable of inherited objects

**Comment by lissyx** [comment]
> @lissyx @wsfulton could you take a look at this?
> 
> Without the SWIG_TypeCast I am getting calls to the wrong vtable of inherited objects

I'm sorry i cant help you.

**Comment by wsfulton** [comment]
Sorry, I'm not that familiar with Javascript the implementation. Does everything work okay without the `%nodefaultctor`?

**Comment by wsfulton** [comment]
The SWIG_TypeCast is used in the other scripting language modules implementation of SWIG_ConvertPtr, so yes, it does look like it is missing. Am surprised it is missing if it is needed for inheritance to work although it may just be required for multiple inheritance rather than single inheritance. What Javascript code causes the bug?

**Comment by murillo128** [comment]
The code that failed was using multiple virtual inheritance indeed. Not
sure if it will fail on other cases too.

**Comment by wsfulton** [comment]
What is the Javascript code that causes it to fail? I'd like to incorporate it into a testcase before accepting the patch.

**Comment by murillo128** [comment]
I think that the issue was something like this:

```
class A 
{
public:
 virtual void bad() = 0;
};

class B 
{
public:
   virtual void good() = 0;
}

class C : public A, B
{
public:
   virtual void bad() { assert(false); };
   virtual void good() {};
};

class D:
{
  void test(B* b) { b.good(); }
};

```

Then the javascript

```
 const c = new C();
 const d = new D();
  d.test(c);
````

The `d.test(c)` calls `A::bad()` instead of `B::good()`  on `c`.

**Comment by wsfulton** [comment]
Fixed now by #1963. I edited your code above and verified it passes now.
